File: python/src/wallet_observer/observer.py
"""Portfolio observation with caller-defined providers and logical identifiers."""

import os
import re
import time
from contextlib import AsyncExitStack
from decimal import Decimal

from ._validation import ObserverError, amount, calculate, integer, unique
from .evm import EvmBalanceProvider
from .provider import (
    BalanceProvider,
    BalanceRequest,
    ProviderFailure,
    StaticValuationProvider,
    ValuationProvider,
)
from .types import BalanceRecord, ObserverConfig, PortfolioSnapshot, WalletNav


def build_snapshot(observed_at_ms: int, balances: list[BalanceRecord]) -> PortfolioSnapshot:
    integer(observed_at_ms)
    seen = set()
    valued: dict[str, Decimal] = {}
    unvalued: dict[str, list[str]] = {}
    for balance in balances:
        key = (balance.wallet_id, balance.asset_id)
        if key in seen:
            raise ObserverError("duplicate_balance")
        seen.add(key)
        for label in (balance.wallet_id, balance.asset_id, balance.chain_id, balance.source):
            unique([label])
        if not re.fullmatch(r"[0-9]{1,39}", balance.atomic_units):
            raise ObserverError("invalid_provider_quantity")
        integer(int(balance.atomic_units), 0, 2**128 - 1)
        expected = (
            None
            if balance.unit_price_usd is None
            else calculate(balance.quantity, "*", balance.unit_price_usd)
        )
        if expected != balance.value_usd:
            raise ObserverError("inconsistent_valuation")
        wallet = balance.wallet_id
        valued.setdefault(wallet, Decimal(0))
        unvalued.setdefault(wallet, [])
        if balance.value_usd is None:
            unvalued[wallet].append(balance.asset_id)
        else:
            valued[wallet] = calculate(valued[wallet], "+", balance.value_usd)
    nav = tuple(WalletNav(w, valued[w], tuple(unvalued[w])) for w in sorted(valued))
    total = Decimal(0)
    for wallet in nav:
        total = calculate(total, "+", wallet.valued_nav_usd)
    return PortfolioSnapshot(1, observed_at_ms, tuple(balances), nav, total)


async def observe_with_providers(
    requests: list[BalanceRequest],
    providers: dict[str, BalanceProvider],
    valuation: ValuationProvider,
    observed_at_ms: int,
) -> PortfolioSnapshot:
    integer(observed_at_ms)
    seen = set()
    for request in requests:
        for label in (request.provider_id, request.wallet_id, request.chain_id, request.asset_id):
            unique([label])
        if (request.wallet_id, request.asset_id) in seen:
            raise ObserverError("duplicate_balance")
        seen.add((request.wallet_id, request.asset_id))
        if request.provider_id not in providers:
            raise ObserverError("unknown_provider")
        integer(request.decimals, 0, 28)
    balances = []
    for request in requests:
        try:
            observed = await providers[request.provider_id].balance(request)
        except ProviderFailure as error:
            raise ProviderFailure(error.code) from None
        except Exception:
            raise ProviderFailure() from None
        try:
            atomic = integer(observed.atomic_units, 0, 2**128 - 1)
            quantity = amount(observed.quantity)
        except (ObserverError, AttributeError):
            raise ProviderFailure("invalid_provider_quantity") from None
        try:
            price = await valuation.unit_price_usd(request)
        except ProviderFailure as error:
            raise ProviderFailure(error.code) from None
        except Exception:
            raise ProviderFailure() from None
        try:
            price = None if price is None else amount(price)
        except ObserverError:
            raise ProviderFailure("invalid_valuation_price") from None
        value = None if price is None else calculate(quantity, "*", price)
        balances.append(
            BalanceRecord(
                request.wallet_id,
                request.chain_id,
                request.asset_id,
                request.symbol,
                str(atomic),
                quantity,
                price,
                value,
                request.provider_id,
            )
        )
    balances.sort(key=lambda b: (b.wallet_id, b.asset_id))
    return build_snapshot(observed_at_ms, balances)


def _required_env(name: str) -> str:
    if not re.fullmatch(r"[A-Za-z_][A-Za-z0-9_]*", name):
        raise ObserverError("invalid_environment_name")
    value = os.environ.get(name)
    if value is None or not value.strip():
        raise ObserverError("missing_environment")
    return value


def _validate_config(config: ObserverConfig):
    if config.schema_version != 1:
        raise ObserverError("schema_version")
    for group in (config.chains, config.wallets, config.assets):
        unique([item.id for item in group])
    chains = {c.id for c in config.chains}
    for item in (*config.wallets, *config.assets):
        if item.chain_id not in chains:
            raise ObserverError("unknown_chain")
    for chain in config.chains:
        if chain.expected_chain_id is not None:
            integer(chain.expected_chain_id, 0, 2**64 - 1)
    for asset in config.assets:
        integer(asset.decimals, 0, 28)
        if asset.kind.type == "native":
            if asset.kind.contract_address_env is not None:
                raise ObserverError("invalid_asset_kind")
        elif asset.kind.type != "erc20" or not asset.kind.contract_address_env:
            raise ObserverError("invalid_asset_kind")


async def observe(config: ObserverConfig) -> PortfolioSnapshot:
    """Collect through owned EVM clients; always close them, including on cancellation."""
    _validate_config(config)
    observed_at_ms = time.time_ns() // 1_000_000
    endpoints = {c.id: _required_env(c.rpc_url_env) for c in config.chains}
    requests = []
    for wallet in config.wallets:
        account = _required_env(wallet.address_env)
        for asset in config.assets:
            if asset.chain_id == wallet.chain_id:
                reference = (
                    _required_env(asset.kind.contract_address_env)
                    if asset.kind.type == "erc20"
                    else None
                )
                requests.append(
                    BalanceRequest(
                        wallet.chain_id,
                        wallet.id,
                        wallet.chain_id,
                        asset.id,
                        asset.symbol,
                        account,
                        reference,
                        asset.decimals,
                    )
                )
    prices = {
        (a.chain_id, a.id): a.valuation_price_usd
        for a in config.assets
        if a.valuation_price_usd is not None
    }
    async with AsyncExitStack() as stack:
        providers = {}
        for chain in config.chains:
            provider = await stack.enter_async_context(EvmBalanceProvider(endpoints[chain.id]))
            if chain.expected_chain_id is not None:
                await provider.verify_chain_id(chain.expected_chain_id)
            providers[chain.id] = provider
        return await observe_with_providers(
            requests,
            providers,
            StaticValuationProvider(prices),
            observed_at_ms,
        )

File: src/observer.rs
use crate::evm::RpcError;
use crate::provider::{
    BalanceProvider, BalanceRequest, EvmBalanceProvider, StaticValuationProvider, ValuationProvider,
};
use crate::types::{AssetKind, BalanceRecord, ObserverConfig, PortfolioSnapshot, WalletNav};
use rust_decimal::Decimal;
use std::collections::{BTreeMap, HashMap, HashSet};
use std::env;
use std::str::FromStr;
use std::sync::Arc;
use std::time::{SystemTime, UNIX_EPOCH};
use thiserror::Error;

#[derive(Debug, Error)]
pub enum ObserverError {
    #[error("configuration schema_version must be 1")]
    SchemaVersion,
    #[error("duplicate identifier: {0}")]
    DuplicateId(String),
    #[error("unknown chain: {0}")]
    UnknownChain(String),
    #[error("missing environment variable: {0}")]
    MissingEnvironment(String),
    #[error("invalid valuation price for {0}")]
    InvalidPrice(String),
    #[error("unknown balance provider: {0}")]
    UnknownProvider(String),
    #[error("balance provider {provider_id} failed: {code}")]
    Provider { provider_id: String, code: String },
    #[error("valuation provider failed: {0}")]
    Valuation(String),
    #[error("decimal overflow")]
    DecimalOverflow,
    #[error(transparent)]
    Rpc(#[from] RpcError),
}

pub async fn observe(config: &ObserverConfig) -> Result<PortfolioSnapshot, ObserverError> {
    validate_config(config)?;
    let observed_at_ms = unix_time_ms();
    let mut providers = HashMap::<String, Arc<dyn BalanceProvider>>::new();
    for chain in &config.chains {
        let endpoint = required_env(&chain.rpc_url_env)?;
        let provider = EvmBalanceProvider::new(endpoint)?;
        if let Some(expected) = chain.expected_chain_id {
            provider
                .verify_chain_id(expected)
                .await
                .map_err(|error| provider_error(&chain.id, error.code))?;
        }
        providers.insert(chain.id.clone(), Arc::new(provider));
    }

    let mut prices = HashMap::new();
    for asset in &config.assets {
        if let Some(raw_price) = &asset.valuation_price_usd {
            let price = Decimal::from_str(raw_price)
                .map_err(|_| ObserverError::InvalidPrice(asset.id.clone()))?;
            if price.is_sign_negative() {
                return Err(ObserverError::InvalidPrice(asset.id.clone()));
            }
            prices.insert((asset.chain_id.clone(), asset.id.clone()), price);
        }
    }

    let mut requests = Vec::new();
    for wallet in &config.wallets {
        let account_reference = required_env(&wallet.address_env)?;
        for asset in config
            .assets
            .iter()
            .filter(|asset| asset.chain_id == wallet.chain_id)
        {
            let asset_reference = match &asset.kind {
                AssetKind::Native => None,
                AssetKind::Erc20 {
                    contract_address_env,
                } => Some(required_env(contract_address_env)?),
            };
            requests.push(BalanceRequest {
                provider_id: wallet.chain_id.clone(),
                wallet_id: wallet.id.clone(),
                chain_id: wallet.chain_id.clone(),
                asset_id: asset.id.clone(),
                symbol: asset.symbol.clone(),
                account_reference: account_reference.clone(),
                asset_reference,
                decimals: asset.decimals,
            });
        }
    }

    let valuation = StaticValuationProvider::new(prices);
    observe_with_providers(&requests, &providers, &valuation, observed_at_ms).await
}

pub async fn observe_with_providers(
    requests: &[BalanceRequest],
    providers: &HashMap<String, Arc<dyn BalanceProvider>>,
    valuation: &dyn ValuationProvider,
    observed_at_ms: i64,
) -> Result<PortfolioSnapshot, ObserverError> {
    let mut balances = Vec::with_capacity(requests.len());
    let mut seen = HashSet::new();
    for request in requests {
        let key = (request.wallet_id.as_str(), request.asset_id.as_str());
        if !seen.insert(key) {
            return Err(ObserverError::DuplicateId(format!(
                "{}/{}",
                request.wallet_id, request.asset_id
            )));
        }
        let provider = providers
            .get(&request.provider_id)
            .ok_or_else(|| ObserverError::UnknownProvider(request.provider_id.clone()))?;
        let observed = provider
            .balance(request)
            .await
            .map_err(|error| provider_error(&request.provider_id, error.code))?;
        if observed.quantity.is_sign_negative() {
            return Err(provider_error(
                &request.provider_id,
                "invalid_provider_quantity".to_string(),
            ));
        }
        let unit_price_usd = valuation
            .unit_price_usd(request)
            .await
            .map_err(|error| ObserverError::Valuation(safe_error_code(&error.code)))?;
        if unit_price_usd.is_some_and(|price| price.is_sign_negative()) {
            return Err(ObserverError::Valuation(
                "invalid_valuation_price".to_string(),
            ));
        }
        let value_usd = unit_price_usd
            .map(|price| {
                observed
                    .quantity
                    .checked_mul(price)
                    .ok_or(ObserverError::DecimalOverflow)
            })
            .transpose()?;
        balances.push(BalanceRecord {
            wallet_id: request.wallet_id.clone(),
            chain_id: request.chain_id.clone(),
            asset_id: request.asset_id.clone(),
            symbol: request.symbol.clone(),
            atomic_units: observed.atomic_units.to_string(),
            quantity: observed.quantity,
            unit_price_usd,
            value_usd,
            source: request.provider_id.clone(),
        });
    }
    balances.sort_by(|left, right| {
        (&left.wallet_id, &left.asset_id).cmp(&(&right.wallet_id, &right.asset_id))
    });
    Ok(build_snapshot(observed_at_ms, balances))
}

pub fn build_snapshot(observed_at_ms: i64, balances: Vec<BalanceRecord>) -> PortfolioSnapshot {
    let mut valued = BTreeMap::<String, Decimal>::new();
    let mut unvalued = BTreeMap::<String, Vec<String>>::new();
    for balance in &balances {
        if let Some(value) = balance.value_usd {
            *valued.entry(balance.wallet_id.clone()).or_default() += value;
        } else {
            unvalued
                .entry(balance.wallet_id.clone())
                .or_default()
                .push(balance.asset_id.clone());
        }
    }
    let wallet_ids: BTreeMap<_, _> = balances
        .iter()
        .map(|balance| (balance.wallet_id.clone(), ()))
        .collect();
    let wallet_nav: Vec<_> = wallet_ids
        .into_keys()
        .map(|wallet_id| WalletNav {
            valued_nav_usd: valued.get(&wallet_id).copied().unwrap_or_default(),
            unvalued_assets: unvalued.remove(&wallet_id).unwrap_or_default(),
            wallet_id,
        })
        .collect();
    let total_valued_nav_usd = wallet_nav.iter().map(|wallet| wallet.valued_nav_usd).sum();
    PortfolioSnapshot {
        schema_version: 1,
        observed_at_ms,
        balances,
        wallet_nav,
        total_valued_nav_usd,
    }
}

fn validate_config(config: &ObserverConfig) -> Result<(), ObserverError> {
    if config.schema_version != 1 {
        return Err(ObserverError::SchemaVersion);
    }
    unique(config.chains.iter().map(|item| item.id.as_str()))?;
    unique(config.wallets.iter().map(|item| item.id.as_str()))?;
    unique(config.assets.iter().map(|item| item.id.as_str()))?;
    let chains: HashSet<_> = config.chains.iter().map(|item| item.id.as_str()).collect();
    for chain_id in config
        .wallets
        .iter()
        .map(|item| item.chain_id.as_str())
        .chain(config.assets.iter().map(|item| item.chain_id.as_str()))
    {
        if !chains.contains(chain_id) {
            return Err(ObserverError::UnknownChain(chain_id.to_string()));
        }
    }
    Ok(())
}

fn unique<'a>(values: impl Iterator<Item = &'a str>) -> Result<(), ObserverError> {
    let mut seen = HashSet::new();
    for value in values {
        if value.is_empty() || !seen.insert(value) {
            return Err(ObserverError::DuplicateId(value.to_string()));
        }
    }
    Ok(())
}

fn required_env(name: &str) -> Result<String, ObserverError> {
    env::var(name)
        .ok()
        .filter(|value| !value.trim().is_empty())
        .ok_or_else(|| ObserverError::MissingEnvironment(name.to_string()))
}

fn provider_error(provider_id: &str, code: String) -> ObserverError {
    ObserverError::Provider {
        provider_id: provider_id.to_string(),
        code: safe_error_code(&code),
    }
}

fn safe_error_code(code: &str) -> String {
    if [
        "chain_id_mismatch",
        "invalid_account_or_asset_reference",
        "invalid_provider_endpoint",
        "invalid_provider_quantity",
        "invalid_valuation_price",
        "provider_failure",
        "provider_method_not_allowed",
        "provider_network_error",
        "provider_response_error",
        "unsupported_asset_decimals",
        "valuation_unavailable",
    ]
    .contains(&code)
    {
        code.to_string()
    } else {
        "provider_failure".to_string()
    }
}

fn unix_time_ms() -> i64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_default()
        .as_millis()
        .min(i64::MAX as u128) as i64
}
